Sources/MongoSwift/BSON/BSONEncoder.swift

```diff
@@ -733,7 +744,16 @@ extension _BSONEncoder: SingleValueEncodingContainer {
 private class MutableArray: BSONValue {
     fileprivate static var bsonType: BSONType { .array }
 
-    fileprivate var bson: BSON { .array(self.array.map { $0.bson }) }
+    fileprivate var bson: BSON {
```

Comment:
so I actually think we should just get rid of these `.bson` properties' implementations altogether and `fatalError` when you try to use them. and we can just special case both of these and have the logic in `toDocument`. maybe you could just move this into a like `toBSONArray()` throwing method on this type? then your `toDocument()` logic becomes something like

```swift
for i in 0..<self.keys.count {
    let value = self.values[i]
    switch value {
    case let val as MutableDictionary:
        try doc.setValue(for: self.keys[i], to: val.toDocument().bson)
    case let val as MutableArray:
        let array = try val.toBSONArray()
        try doc.setValue(for: self.key[i], to: array.bson)
    default:
        try doc.setValue(for: self.keys[i], to: value.bson)
    }
}
```